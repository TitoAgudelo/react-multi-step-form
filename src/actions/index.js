import {
  STEP_ONE, STEP_TWO, STEP_THREE, STEP_FOUR,
  SUBMIT_FORM_DATA, SUBMIT_FORM_DATA_SUCCESS, SUBMIT_FORM_DATA_ERROR,
  CHECK_TEXT, CHECK_TEXT_SUCCESS, CHECK_TEXT_FAILURE
} from './../constants/';

import { checkIt, submitIt } from './../api';

export function stepOne(formInfo) {
  return {
    type: STEP_ONE,
    payload: formInfo,
  };
}

export function stepTwo(formInfo) {
  return {
    type: STEP_TWO,
    payload: formInfo,
  };
}

function submitData() {
  return {
    type: SUBMIT_FORM_DATA,
  };
}

function submitDataSuccess() {
  return {
    type: SUBMIT_FORM_DATA_SUCCESS,
  };
}

function submitDataFailure(error) {
  return {
    type: SUBMIT_FORM_DATA_ERROR,
    payload: {
      error,
    },
  };
}

function checkText() {
  return {
    type: CHECK_TEXT,
  };
}

function checkTextSuccess() {
  return {
    type: CHECK_TEXT_SUCCESS,
  };
}

function checkTextFailure(error) {
  return {
    type: CHECK_TEXT_FAILURE,
    payload: {
      error,
    },
  };
}

export function stepThree(formInfo) {
  return (dispatch) => {
    dispatch(checkText());
    checkIt(formInfo.value)
    .then(() => {
      dispatch(checkTextSuccess());
      dispatch({
        type: STEP_THREE,
        payload: formInfo,
      });
    }, (error) => {
      dispatch(checkTextFailure(error.message));
    });
  };
}


export function stepFour(formInfo) {
  return {
    type: STEP_FOUR,
    payload: formInfo,
  };
}

export function stepFive() {
  return (dispatch, getState) => {
    const state = getState();
    let formPayload = state.form.payload;
    dispatch(submitData());
    submitIt(formPayload).then(() => {
        dispatch(submitDataSuccess());
      }, (error) => {
        dispatch(submitDataFailure(error.message));
      });
  };
}
