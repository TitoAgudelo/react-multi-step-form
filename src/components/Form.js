import React from 'react'
import classnames from 'classnames'
import { PropTypes } from 'prop-types';

import {
  PENDING, FAILED
} from './../constants/'

const options = ['C1', 'C2', 'C3']

export default class StepForm extends React.Component {

  static propTypes = {
    form: PropTypes.object,
    actions: PropTypes.object
  };

  constructor(props) {
    super(props);
    this.state = {
      currentStep: props.form.nextStep,
      A1: false,
      A2: false,
      B1: false,
      B2: false,
      step3Text: '',
      selectedOption: options[0]
    };
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      currentStep: nextProps.form.nextStep,
      checkTextStatus: nextProps.form.checkTextStatus,
      formSubmissionStatus: nextProps.form.formSubmissionStatus,
      formSubmissionError: nextProps.form.formSubmissionError,
      checkTextError: nextProps.form.checkTextError
    });
  }

  renderErrorText(error) {
    let errorText = error || 'Error: Please try again.';
    return (
     <div>
       <span className="error-text">{errorText}</span>
     </div>
    );
  }

  /*
    Step 1
   */

  onCheckBoxesClicked(event) {
    let checkBoxId = event.target.id;
    let isA1Checked = checkBoxId === 'A1' ? !this.state.A1 : this.state.A1;
    let isA2Checked = checkBoxId === 'A2' ? !this.state.A2 : this.state.A2;
    this.setState({
      A1: isA1Checked,
      A2: isA2Checked
    });
    this.props.actions.stepOne({
      A1: isA1Checked,
      A2: isA2Checked
    });
  }

  renderStepOne() {
    return (
      <div className="step1">
        <h3>Step one</h3>
        <input type="checkbox" id="A1" value={this.state.A1} onChange={this.onCheckBoxesClicked.bind(this)} />
        <label> A1</label>
        <br/>
        <input type="checkbox" id="A2" value={this.state.A2} onChange={this.onCheckBoxesClicked.bind(this)} />
        <label>A2</label>
      </div>
    );
  }

  /*
    Step 2
   */

  onStepTwoButtonsClicked(event) {
    let buttonId = event.target.id;
    let b2Selected = buttonId === 'B2' ? true : false;
    let b1Selected = buttonId === 'B1' ? true : false;
    this.setState({
      B2: b2Selected,
      B1: b1Selected
    });
    this.props.actions.stepTwo({
      B2: b2Selected,
      B1: b1Selected
    });
  }

  renderStepTwo() {
    let b1ClassNames = classnames({
      'button-step-two': true,
      'active':  this.state.B1 === true
    });

    let b2ClassNames = classnames({
      'button-step-two': true,
      'active': this.state.B2 === true
    });

    return (
      <div className="step2">
        <h3>Step two</h3>
        <div className={b1ClassNames} id="B1" onClick={this.onStepTwoButtonsClicked.bind(this)}>
          B1
        </div>
        <div className={b2ClassNames} id="B2" onClick={this.onStepTwoButtonsClicked.bind(this)}>
          B2
        </div>
      </div>
    );
  }

  /*
    Step 3
   */

  handleStep3Change(event) {
    this.setState({step3Text: event.target.value});
  }

  checkStepThree() {
    this.props.actions.stepThree({
      value: this.state.step3Text
    });
  }

  renderStepThree() {
    let checkButtonClasses = classnames({
      'button-step-two': true,
      'active':  this.state.checkTextStatus !== PENDING
    });
    return (
      <div className="step3">
        <h3>Step three</h3>
        <input type="text" value={this.state.step3Text} onChange={this.handleStep3Change.bind(this)} />
        <div className={checkButtonClasses} onClick={this.checkStepThree.bind(this)}>
          Check
        </div>
        {this.state.checkTextStatus === FAILED ? this.renderErrorText(this.state.checkTextError) : ''}
      </div>
    );
  }

  /*
    Step 4
   */

  onSelectChange(event) {
    let option = event.target.value;
    this.setState({selectedOption: option});
    this.props.actions.stepFour({
      value: option
    });
  }

  renderOptions(options) {
    return (
      options.map((option) => {
        return (
          <option key={option} value={option}>
            {option}
          </option>
        );
      })
    );
  }

  renderStepFour() {
    return (
      <div className="step4">
        <h3>Step four</h3>
        <select onChange={this.onSelectChange.bind(this)}>
          {
            this.renderOptions(options)
          }
        </select>
      </div>
    );
  }

  /*
    Final step
   */

  renderSubmit() {
    let disabled = this.state.formSubmissionStatus === PENDING;
    let showErorr = this.state.formSubmissionStatus === FAILED;
    if(disabled){
      return (
        <input type="submit" value="Submit" disabled/>
      );
    }
    return (
      <div>
        {showErorr ? this.renderErrorText(this.state.formSubmissionError) : ''}
        <input type="submit" value="Submit" />
      </div>
    );
  }

  handleSubmit(event) {
    event.preventDefault();
    this.props.actions.stepFive();
  }

  /*
    Render the steps
   */
  render() {
    let stepOneJSX = this.state.currentStep  >= 1 ? this.renderStepOne() : '';
    let stepTwoJSX = this.state.currentStep  >= 2 ? this.renderStepTwo() : '';
    let stepThreeJSX = this.state.currentStep  >= 3 ? this.renderStepThree() : '';
    let stepFourJSX = this.state.currentStep  >= 4 ? this.renderStepFour() : '';
    let formButtonJSX = this.state.currentStep >= 5 ? this.renderSubmit() : '';
    return (
      <form onSubmit={this.handleSubmit.bind(this)}>
        <div>
          {stepOneJSX}
          <br/>
          {stepTwoJSX}
          <br/>
          {stepThreeJSX}
          <br/>
          {stepFourJSX}
          <br/>
          {formButtonJSX}
        </div>
      </form>
    );
  }
}
