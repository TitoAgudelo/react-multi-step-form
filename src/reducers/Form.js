import {
  STEP_ONE, STEP_TWO, STEP_THREE, STEP_FOUR,
  SUBMIT_FORM_DATA, SUBMIT_FORM_DATA_SUCCESS, SUBMIT_FORM_DATA_ERROR,
  CHECK_TEXT, CHECK_TEXT_SUCCESS, CHECK_TEXT_FAILURE,
  PENDING, SUCCESS, FAILED
} from './../constants/';

const initialState = {
  nextStep: 1,
  payload: {}
};

export default function reducer(state = initialState, action) {
  switch(action.type) {
    case STEP_ONE:
      let A1 = action.payload.A1 === true ? 'A1' : '';
      let A2 = action.payload.A2 === true ? 'A2' : '';
      let a = [A1, A2].filter((item) => {
        return item !== '';
      });
      return {
        ...state,
        nextStep: 2,
        payload: Object.assign({}, state.payload,{
          a: a
        })
      };

    case STEP_TWO:
      let b = action.payload.B1 === true ? 'B1' : 'B2';
      return {
        ...state,
        nextStep: 3,
        payload: Object.assign({}, state.payload,{
          b: b
        })
      };

    case STEP_THREE:
      return {
        ...state,
        nextStep: 4,
        payload: Object.assign({}, state.payload,{
          text: action.payload.value
        })
      };

    case STEP_FOUR:
      return {
        ...state,
        nextStep: 5,
        payload: Object.assign({}, state.payload,{
          c: action.payload.value
        })
      };

    case SUBMIT_FORM_DATA:
      return {
        ...state,
        formSubmissionStatus: PENDING
      };

    case SUBMIT_FORM_DATA_SUCCESS:
      return {
        ...state,
        formSubmissionStatus: SUCCESS
      };

    case SUBMIT_FORM_DATA_ERROR:
      return {
        ...state,
        formSubmissionStatus: FAILED,
        formSubmissionError: action.payload.error
      };

    case CHECK_TEXT:
      return {
        ...state,
        checkTextStatus: PENDING
      };

    case CHECK_TEXT_SUCCESS:
      return {
        ...state,
        checkTextStatus: SUCCESS
      };

    case CHECK_TEXT_FAILURE:
      return {
        ...state,
        checkTextStatus: FAILED,
        checkTextError: action.payload.error
      };

    default:
      return state;
  }
}
