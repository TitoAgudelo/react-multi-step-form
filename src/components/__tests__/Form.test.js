import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16';
import Form from './../Form';
import * as actions from './../../actions';

it('renders correctly', () => {
  const form = {
    nextStep: 1,
    payload: {},
  };

  const tree = renderer
    .create(
      <Form form={form} actions={actions}/>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
