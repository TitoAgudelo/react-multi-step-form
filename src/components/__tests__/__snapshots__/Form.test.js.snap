// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="container"
>
  <h1
    className="title"
  >
    Multistep React & Redux
  </h1>
  <div
    className="stepline"
  >
    <div
      className="line active"
    >
      step one
    </div>
    <div
      className="line "
    >
      step two
    </div>
    <div
      className="line "
    >
      step three
    </div>
    <div
      className="line "
    >
      step four
    </div>
    <div
      className="line "
    >
      submit
    </div>
  </div>
  <form
    className="form-container"
    onSubmit={[Function]}
  >
    <div>
      <div
        className="step1"
      >
        <h3
          className="step-title"
        >
          Step one
        </h3>
        <p
          className="step-legend"
        >
          Next step is available after at least one of them is checked
        </p>
        <input
          id="A1"
          onChange={[Function]}
          type="checkbox"
          value={false}
        />
        <label>
          A1
        </label>
        <br />
        <input
          id="A2"
          onChange={[Function]}
          type="checkbox"
          value={false}
        />
        <label>
          A2
        </label>
      </div>
      
      
      
      
    </div>
  </form>
</div>
`;
